web/src/main/java/com/marklogic/hub/web/service/AsyncFlowService.java

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright 2012-2019 MarkLogic Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.marklogic.hub.web.service;
+
+import com.marklogic.hub.FlowManager;
+import com.marklogic.hub.flow.Flow;
+import com.marklogic.hub.flow.impl.FlowRunnerImpl;
+import com.marklogic.hub.job.JobStatus;
+import com.marklogic.hub.util.metrics.tracer.JaegerConfig;
+import com.marklogic.hub.web.exception.DataHubException;
+import com.marklogic.hub.web.model.FlowJobModel.FlowJobs;
+import com.marklogic.hub.web.model.FlowStepModel;
+import io.opentracing.Scope;
+import io.opentracing.Span;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.IntStream;
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class AsyncFlowService {
+    protected final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    //core/maximum pool size, can be configurable if required
+    private static final int NTHREADS = 6;
+
+    //a thread processes 10 flows, can be configurable if required
+    //if the flow number is not more than 10, we do not use any thread in the pool
+    private static final int FLOW_COUNT_PER_THREAD = 10;
+
+    @Autowired
+    private FlowManager flowManager;
+
+    @Autowired
+    private FlowRunnerImpl flowRunner;
+
+    @Autowired
+    private FlowJobService flowJobService;
+
+    ExecutorService executor;
+
+    @PostConstruct
+    public void init(){
+        //we could turn on/off jaeger to trace performance of any call stacks by
+        //injecting customized span code along with tags, e.g. try (Scope scope = JaegerConfig.activate(span))
+        //default tracing is off, in order to turn on, just add a VM option -DJaegerServiceName=data-hub
+        //also see: https://www.jaegertracing.io/docs/1.12/getting-started/
+        String jaegerServiceName = System.getProperty("JaegerServiceName");
```

Comment:
Instead of pulling a system prop, I recommend using a Spring Value annotation. That allows for the property to be set in a number of ways, not just via system props.

Reply:
Now Jaeger is only internally used for us to trace the bottleneck of performance and then try to improve it. Once you turn on it with the system property flag, and measure what codes you want to trace by simply injecting code snippet, adding some tags. It will print out detail log information. The most important thing is you can analyze the performance using a clean graphical web interface once you run the Jaeger server locally or remotely (docker image).

Reply:
@rjrudin good suggestion! done.